frb_codegen/src/parser/mod.rs

```diff
@@ -187,20 +187,163 @@ impl<'a> Parser<'a> {
     }
 }
 
-fn extract_fns_from_file(file: &File) -> Vec<&ItemFn> {
+fn extract_fns_from_file(file: &File) -> Vec<ItemFn> {
     let mut src_fns = Vec::new();
 
     for item in file.items.iter() {
         if let Item::Fn(ref item_fn) = item {
             if let Visibility::Public(_) = &item_fn.vis {
-                src_fns.push(item_fn);
+                src_fns.push(item_fn.clone());
+            }
+        }
+        if let Item::Impl(ref item_impl) = item {
+            for item in &item_impl.items {
+                if let ImplItem::Method(item_method) = item {
+                    if let Visibility::Public(_) = &item_method.vis {
+                        let f = item_method_to_function(item_impl, item_method)
+                            .expect("item implementation is unsupported");
+                        src_fns.push(f);
+                    }
+                }
             }
         }
     }
 
     src_fns
 }
 
+// Converts an item implementation (something like fn(&self, ...)) into a function where `&self` is a named parameter to `&Self`
+fn item_method_to_function(item_impl: &ItemImpl, item_method: &ImplItemMethod) -> Option<ItemFn> {
```

Comment:
Well, if I understatnd correctly, this approach is: convert a method into a (fake) function AST, then use `parse_function` to convert that fake function AST into our IR (IrFunc).

Then, what about mimic `parse_function` and create `parse_method` to directly convert ItemImpl to IrFunc? 

Surely, the possibly duplicated code between parse_function and parse_method should be extracted, if any

Reply:
@lattice0 seems this comment is folded so needs to be manually expanded to see it

Reply:
I don't understand what is the problem here. I convert to `ItemFn` because this is what the other function did. I just do it but for ItemImpl. Then later it is converted just like the other function.

Reply:
Well, IMHO it is as follows:

Current approach:

```
syn::ImplItemMethod ---frb:: item_method_to_function---> syn::ItemFn
syn::ItemFn ----frb::parse_function---> frb::IrFunc
```

Proposed approach:

```
syn::ImplItemMethod ---frb::parse_method---> frb::IrFunc
syn::ItemFn ----frb::parse_function---> frb::IrFunc
```

The current approach creates some "artificial" temporary syn::ItemFn which does not seem to be elegant, since we indeed want IrFunc.

However, if it is very hard to convert syn::ImplItemMethod into frb::IrFunc directly (probably with some refactoring on parse_function), then the current approach seems ok.

Reply:
It would require a re-do of lots of things, right?

Reply:
I guess maybe not. But anyway the current approach seems acceptable.